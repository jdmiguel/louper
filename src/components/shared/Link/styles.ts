import { styled } from '@mui/material/styles';
import MuiLink from '@mui/material/Link';
import Icon from '@mui/material/Icon';

export const StyledLink = styled(MuiLink)(({ theme }) => ({
  cursor: 'pointer',
  alignItems: 'center',
  display: 'flex',
  textDecoration: 'none',
  transition: 'color ease-out 250ms',
  span: { transition: 'color ease-out 250ms' },
  '&:hover': {
    color: theme.palette.primary.light,
    span: { color: theme.palette.primary.light },
  },
}));

export const StyledIcon = styled(Icon)(({ theme }) => ({
  color: theme.palette.primary.main,
  fontSize: 22,
  marginRight: 5,
}));
