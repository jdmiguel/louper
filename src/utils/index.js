export const errorLiterals = {
  maximumRequest:
    'You have excedeed the maximum allowed request. Please, wait for a while',
  unavailableUser: 'Please, choose an available user',
};

export const externalLink = (url) => window.open(url, '_blank');
