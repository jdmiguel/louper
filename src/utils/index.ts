import { ResponseError } from './types';

enum ErrorMessage {
  MAX = 'You have excedeed the maximum allowed request. Please, wait for a while',
  NO_USER = 'Please, choose an available user',
  GENERIC = 'Sorry! there was an error on the server side.',
}

export const getErrorMessage = (error: ResponseError) => {
  if (error.status === 403) {
    return ErrorMessage.MAX;
  }

  if (error.status === 404) {
    return ErrorMessage.NO_USER;
  }

  return ErrorMessage.GENERIC;
};

export const debounce = (fn: any, delay: number) => {
  let timer: ReturnType<typeof setTimeout>;

  return (...args: any) => {
    if (timer) clearTimeout(timer);
    timer = setTimeout(() => {
      fn(...args);
    }, delay);
  };
};
