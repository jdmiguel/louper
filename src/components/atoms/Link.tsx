/* material-ui */
import { styled } from '@mui/material/styles';
import MuiLink from '@mui/material/Link';
import Icon from '@mui/material/Icon';

const StyledLink = styled(MuiLink)(({ theme }) => ({
  cursor: 'pointer',
  alignItems: 'center',
  display: 'flex',
  fontSize: '0.9rem',
  fontWeight: 400,
  lineHeight: 0,
  textDecoration: 'none',
  transition: 'color ease-out 250ms',
  '& span': { transition: 'color ease-out 250ms' },
  '&:hover': {
    color: theme.palette.primary.light,
    '& span': { color: theme.palette.primary.light },
  },
}));

const StyledIcon = styled(Icon)(({ theme }) => ({
  color: theme.palette.primary.dark,
  fontSize: 22,
  marginRight: 5,
}));

type LinkProps = {
  url: string;
  ariaLabel: string;
  content: string;
};
type IconProps = { withIcon?: false; iconType?: never } | { withIcon: true; iconType?: string };
type Props = LinkProps & IconProps;

const Link = ({ url, ariaLabel, content, withIcon, iconType }: Props) => (
  <StyledLink href={url} rel="noopener noreferrer" aria-label={ariaLabel}>
    {withIcon && <StyledIcon>{iconType}</StyledIcon>}
    {content}
  </StyledLink>
);

export default Link;
