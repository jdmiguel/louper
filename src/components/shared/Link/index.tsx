import { StyledLink, StyledIcon } from './styles';

type LinkProps = {
  url: string;
  ariaLabel: string;
  content: string;
};
type IconProps = { withIcon?: false; iconType?: never } | { withIcon: true; iconType?: string };
type Props = LinkProps & IconProps;

const Link = ({ url, ariaLabel, content, withIcon, iconType }: Props) => (
  <StyledLink href={url} rel="noopener noreferrer" aria-label={ariaLabel}>
    {withIcon && <StyledIcon>{iconType}</StyledIcon>}
    {content}
  </StyledLink>
);

export default Link;
