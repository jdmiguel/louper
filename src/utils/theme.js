/* material-ui */
import { keyframes } from '@emotion/react';
import { createTheme } from '@mui/material/styles';

export const theme = createTheme({
  overrides: {
    MuiCssBaseline: {
      '@global': {
        html: {
          overflowX: 'hidden',
          background: '#E0E6EF',
        },
        body: {
          margin: 0,
          padding: 0,
        },
        button: {
          boxShadow: 'initial !important',
        },
      },
    },
  },
  typography: {
    fontFamily: ['Comfortaa', 'Arial', 'sans-serif'].join(','),
    htmlFontSize: 12,
    h1: {
      color: '#353D40',
      fontSize: '4.8rem',
      fontWeight: 700,
      letterSpacing: '-0.25rem',
      lineHeight: 0,
      '@media (min-width: 769px)': {
        fontSize: '6rem',
      },
    },
    h2: {
      color: '#353D40',
      fontSize: '3.5rem',
      fontWeight: 300,
      letterSpacing: '-0.1rem',
      lineHeight: 0,
      '@media (min-width: 769px)': {
        fontSize: '4.2rem',
      },
    },
    h3: {
      color: '#353D40',
      fontSize: '1.3rem',
      fontWeight: 700,
      '@media (min-width: 769px)': {
        fontSize: '1.5rem',
      },
    },
    h4: {
      color: '#353D40',
      fontSize: '1.1rem',
      fontWeight: 500,
      '@media (min-width: 769px)': {
        fontSize: '1.2rem',
      },
    },
    h5: {
      color: '#6b77a1',
      fontSize: '0.9rem',
      fontWeight: 700,
      textTransform: 'uppercase',
    },
    h6: {
      color: '#757d81',
      fontSize: '0.8rem',
      fontWeight: 500,
      textTransform: 'uppercase',
    },
    button: {
      fontSize: '0.9rem',
    },
    body1: {
      color: '#353D40',
      fontSize: '0.9rem',
      lineHeight: '1.1rem',
    },
    body2: {
      color: '#E1D7EC',
      fontSize: '0.8rem',
      lineHeight: '1rem',
    },
    subtitle2: {
      color: '#353D40',
      fontSize: '0.8rem',
      fontWeight: 500,
      textTransform: 'uppercase',
    },
  },
  palette: {
    background: {
      default: '#E0E6EF',
    },
    primary: {
      main: '#553285',
      light: '#7B52AB',
      dark: '#36175E',
      contrastText: '#E1D7EC',
    },
    secondary: {
      main: '#6b77a1',
      light: '#c7cde1',
      dark: '#565e7a',
      contrastText: '#bac1d9',
    },
    neutral: {
      main: '#353D40',
    },
  },
  animation: {
    greet: keyframes`
      0%,100% { transform: rotate(0) }
      20%,60% { transform: rotate(-25deg) }
      40%,80% { transform: rotate(10deg) }
    `,
  },
});
