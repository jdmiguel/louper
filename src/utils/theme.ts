/* material-ui */
import { keyframes } from '@emotion/react';
import { createTheme } from '@mui/material/styles';

/* colors */
import { colors } from './colors';

type Animation = {
  greet: any;
  shimmer: any;
};

declare module '@mui/material/styles' {
  interface Theme {
    animation: Animation;
    overrides: any;
  }
  interface ThemeOptions {
    animation: Animation;
    overrides: any;
  }
  interface Palette {
    neutral: Palette['primary'];
    overlay: Palette['primary'];
  }
  interface PaletteOptions {
    neutral: PaletteOptions['primary'];
    overlay: PaletteOptions['primary'];
  }
}

export const theme = createTheme({
  animation: {
    greet: keyframes`
      0%,100% { transform: rotate(0) }
      20%,60% { transform: rotate(-25deg) }
      40%,80% { transform: rotate(10deg) }
    `,
    shimmer: keyframes`
      0% { background-position-x: -1200px}
      100% { background-position-x: 1200px}
    `,
  },
  overrides: [
    {
      MuiCssBaseline: {
        '@global': {
          html: {
            overflowX: 'hidden',
            background: colors.darkBlue,
          },
          body: {
            margin: 0,
            padding: 0,
          },
          button: {
            boxShadow: 'initial !important',
          },
        },
      },
    },
  ],
  typography: {
    fontFamily: ['Comfortaa', 'Arial', 'sans-serif'].join(','),
    htmlFontSize: 12,
    h2: {
      fontSize: '1.1rem',
      fontWeight: 300,
      '@media (min-width: 768px)': {
        fontSize: '1.3rem',
      },
    },
    h3: {
      fontSize: '1.3rem',
      fontWeight: 700,
      '@media (min-width: 768px)': {
        fontSize: '1.5rem',
      },
    },
    h4: {
      fontSize: '1.1rem',
      fontWeight: 500,
      '@media (min-width: 768px)': {
        fontSize: '1.2rem',
      },
    },
    h5: {
      fontSize: '1rem',
      fontWeight: 700,
      textTransform: 'uppercase',
    },
    h6: {
      fontSize: '0.9rem',
      fontWeight: 700,
      textTransform: 'uppercase',
    },
    body1: {
      fontSize: '1em',
    },
    body2: {
      fontSize: '0.9em',
    },
    overline: {
      fontSize: '0.8rem',
      fontWeight: 700,
      lineHeight: '1rem',
      textTransform: 'uppercase',
    },
  },
  palette: {
    primary: {
      main: colors.pink,
      light: colors.lightPink,
      dark: colors.darkPink,
    },
    secondary: {
      main: colors.blue,
      light: colors.lightBlue,
      dark: colors.darkBlue,
    },
    neutral: {
      main: colors.grey,
      light: colors.lightGrey,
    },
    overlay: {
      main: colors.darkPinkOverlay,
    },
    background: {
      default: colors.darkBlue,
    },
    text: {
      primary: colors.white,
      secondary: colors.lightBlue,
    },
  },
});
