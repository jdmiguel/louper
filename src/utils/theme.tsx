import { ReactElement } from 'react';
import { keyframes } from '@emotion/react';
import { Keyframes } from '@emotion/serialize';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { colors } from './colors';

type Animation = {
  greet: Keyframes;
  fadeIn: Keyframes;
  fadeInUp: Keyframes;
  fadeInLeft: Keyframes;
  fadeInRight: Keyframes;
};

declare module '@mui/material/styles' {
  interface Theme {
    animation: Animation;
  }
  interface ThemeOptions {
    animation: Animation;
  }
  interface Palette {
    neutral: Palette['primary'];
    overlay: Palette['primary'];
  }
  interface PaletteOptions {
    neutral: PaletteOptions['primary'];
    overlay: PaletteOptions['primary'];
  }
}

export const theme = createTheme({
  animation: {
    greet: keyframes`
      0%,100% { transform: rotate(0) }
      20%,60% { transform: rotate(-25deg) }
      40%,80% { transform: rotate(10deg) }
    `,
    fadeIn: keyframes`
      0% { opacity:0 }
      100% { opacity:1 }
    `,
    fadeInUp: keyframes`
      0% { opacity:0; transform: translateY(20px); }
      100% { opacity:1; transform: translateY(0px);  }
    `,
    fadeInLeft: keyframes`
      0% { opacity:0; transform: translateX(-50px); }
      100% { opacity:1; transform: translateX(0px);  }
    `,
    fadeInRight: keyframes`
      0% { opacity:0; transform: translateX(50px); }
      100% { opacity:1; transform: translateX(0px);  }
    `,
  },
  components: {
    MuiCssBaseline: {
      styleOverrides: {
        body: {
          overflowX: 'hidden',
        },
        h1: {
          margin: 0,
          lineHeight: 0,
        },
        h2: {
          margin: 0,
          lineHeight: 0,
        },
      },
    },
  },
  typography: {
    fontFamily: ['Comfortaa', 'Arial', 'sans-serif'].join(','),
    htmlFontSize: 12,
    h2: {
      fontSize: '0.9rem',
      '@media (min-width: 480px)': {
        fontSize: '1rem',
      },
      '@media (min-width: 768px)': {
        fontSize: '1.1rem',
      },
      '@media (min-width: 1440px)': {
        fontSize: '1.15rem',
      },
      fontWeight: 300,
    },
    h3: {
      fontSize: '1.3rem',
      fontWeight: 700,
      '@media (min-width: 768px)': {
        fontSize: '1.5rem',
      },
    },
    h4: {
      fontSize: '1.1rem',
      fontWeight: 500,
      '@media (min-width: 768px)': {
        fontSize: '1.2rem',
      },
    },
    h5: {
      fontSize: '1rem',
      fontWeight: 700,
      textTransform: 'uppercase',
    },
    h6: {
      fontSize: '0.9rem',
      fontWeight: 700,
      textTransform: 'uppercase',
    },
    body1: {
      fontSize: '1rem',
    },
    body2: {
      fontSize: '0.9rem',
    },
    overline: {
      fontSize: '0.8rem',
      fontWeight: 700,
      lineHeight: '1rem',
      textTransform: 'uppercase',
    },
  },
  palette: {
    primary: {
      main: colors.pink,
      light: colors.lightPink,
      dark: colors.darkPink,
    },
    secondary: {
      main: colors.blue,
      light: colors.lightBlue,
      dark: colors.darkBlue,
    },
    error: {
      main: colors.red,
    },
    neutral: {
      dark: colors.grey,
      main: colors.lightGrey,
      light: colors.white,
    },
    overlay: {
      main: colors.darkPinkOverlay,
    },
    background: {
      default: colors.darkBlue,
      paper: colors.gradient,
    },
    text: {
      primary: colors.white,
      secondary: colors.lightBlue,
    },
  },
});

export const renderWithTheme = (component: ReactElement) => (
  <ThemeProvider theme={theme}>{component}</ThemeProvider>
);
