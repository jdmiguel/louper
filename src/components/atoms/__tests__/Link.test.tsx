import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { renderWithTheme } from '../../../utils/theme';
import Link from '../Link';

describe('<Link />', () => {
  const props = {
    url: 'https://github.com/JohnDoe',
    ariaLabel: 'View John Doe profile on GitHub',
    content: 'VISIT PROFILE',
  };

  it('renders the content properly', () => {
    const { container } = render(renderWithTheme(<Link {...props} />));
    expect(container).toMatchSnapshot();
  });

  it('renders the content properly when there is an icon associated', () => {
    const { container } = render(
      renderWithTheme(<Link {...props} withIcon iconType="visibility" />),
    );
    expect(container).toMatchSnapshot();
  });
});
