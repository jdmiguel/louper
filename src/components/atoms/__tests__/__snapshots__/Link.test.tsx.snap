// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> renders the content properly 1`] = `
<div>
  <a
    aria-label="View John Doe profile on GitHub"
    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-vnlgj6-MuiTypography-root-MuiLink-root"
    href="https://github.com/JohnDoe"
    rel="noopener noreferrer"
  >
    VISIT PROFILE
  </a>
</div>
`;

exports[`<Link /> renders the content properly when there is an icon associated 1`] = `
<div>
  <a
    aria-label="View John Doe profile on GitHub"
    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-vnlgj6-MuiTypography-root-MuiLink-root"
    href="https://github.com/JohnDoe"
    rel="noopener noreferrer"
  >
    <span
      aria-hidden="true"
      class="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-l9i5fy-MuiIcon-root"
    >
      visibility
    </span>
    VISIT PROFILE
  </a>
</div>
`;
